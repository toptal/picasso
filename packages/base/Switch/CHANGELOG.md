# @toptal/picasso-switch

## 1.0.6

### Patch Changes

- Updated dependencies [[`4ee1ebd`](https://github.com/toptal/picasso/commit/4ee1ebdafd9e5830d5ec6007620186d5a61befee)]:
  - @toptal/picasso-form@2.0.0

## 1.0.5

### Patch Changes

- Updated dependencies [[`c05387d`](https://github.com/toptal/picasso/commit/c05387de5636ed094365d1eff67b955d84b81c61)]:
  - @toptal/picasso-form@1.1.3

## 1.0.4

### Patch Changes

- Updated dependencies [[`3512588`](https://github.com/toptal/picasso/commit/3512588b06c3660471a68500275321c640278cf0)]:
  - @toptal/picasso-form@1.1.2

## 1.0.3

### Patch Changes

- Updated dependencies [[`a1d523092cc4de6cb376156435b99b1e483f39b9`](https://github.com/toptal/picasso/commit/a1d523092cc4de6cb376156435b99b1e483f39b9)]:
  - @toptal/picasso-shared@14.0.1
  - @toptal/picasso-form@1.1.1

## 1.0.2

### Patch Changes

- Updated dependencies [[`8142d393588c8c3cf5193b4273957646258883e2`](https://github.com/toptal/picasso/commit/8142d393588c8c3cf5193b4273957646258883e2)]:
  - @toptal/picasso-form@1.1.0

## 1.0.1

### Patch Changes

- [#4164](https://github.com/toptal/picasso/pull/4164) [`84d579cee65c88aeb6ba58318c43d8deb5201b30`](https://github.com/toptal/picasso/commit/84d579cee65c88aeb6ba58318c43d8deb5201b30) Thanks [@mkrl](https://github.com/mkrl)!
- migrate to Picasso by Parts
  - picasso is now distributed as a set of independent packages with the main package `@toptal/picasso` now being a collection of re-exported packages
- Updated dependencies [[`84d579cee65c88aeb6ba58318c43d8deb5201b30`](https://github.com/toptal/picasso/commit/84d579cee65c88aeb6ba58318c43d8deb5201b30)]:
  - @toptal/picasso-shared@14.0.0
  - @toptal/picasso-form@1.0.1
