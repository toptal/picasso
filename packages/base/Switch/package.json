{
  "name": "@toptal/picasso-switch",
  "version": "2.0.1",
  "description": "Toptal UI components library - Switch",
  "publishConfig": {
    "access": "public"
  },
  "main": "./dist-package/src/index.js",
  "module": "./dist-package/src/index.js",
  "scripts": {
    "build:package": "tsc -b tsconfig.json",
    "prepublishOnly": "yarn build:package"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/toptal/picasso.git"
  },
  "author": "Toptal",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/toptal/picasso/issues"
  },
  "homepage": "https://github.com/toptal/picasso/tree/master/packages/picasso#readme",
  "dependencies": {
    "@toptal/picasso-form": "3.0.5",
    "@toptal/picasso-shared": "15.0.0",
    "@mui/base": "5.0.0-beta.40",
    "classnames": "^2.5.1"
  },
  "sideEffects": [
    "**/styles.ts",
    "**/styles.js"
  ],
  "peerDependencies": {
    "@toptal/picasso-provider": "*",
    "@toptal/picasso-tailwind": "^2.5.0",
    "react": ">=16.12.0 < 19.0.0"
  },
  "exports": {
    ".": "./dist-package/src/index.js"
  },
  "devDependencies": {
    "@toptal/picasso-provider": "4.2.1",
    "@toptal/picasso-test-utils": "1.1.1"
  },
  "files": [
    "dist-package/**",
    "!dist-package/tsconfig.tsbuildinfo",
    "src"
  ]
}
