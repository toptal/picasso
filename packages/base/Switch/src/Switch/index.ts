import type { OmitInternalProps } from '@toptal/picasso-shared'

import type { Props as InternalSwitchProps } from './Switch'
export { default as Switch } from './Switch'
export type SwitchProps = OmitInternalProps<InternalSwitchProps>
