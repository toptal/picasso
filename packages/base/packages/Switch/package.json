{
  "name": "@toptal/picasso-switch",
  "version": "1.0.0",
  "main": "index.js",
  "dependencies": {
    "@toptal/picasso-form-control-label": "^1.0.0"
  }
}