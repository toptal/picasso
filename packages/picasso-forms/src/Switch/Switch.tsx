import React from 'react'
import type { SwitchProps } from '@toptal/picasso'
import { Switch as PicassoSwitch, Form as PicassoForm } from '@toptal/picasso'

import type { FieldProps } from '../Field'
import PicassoField from '../Field'
import type { Props as FieldLabelProps } from '../FieldLabel'

export type FormSwitchProps = Omit<SwitchProps, 'onChange'> & {
  onChange?: SwitchProps['onChange']
}
export type Props = FormSwitchProps &
  FieldProps<SwitchProps['value']> &
  Omit<FieldLabelProps, 'name' | 'required'>

export const Switch = (props: Props) => (
  <PicassoField<FormSwitchProps>
    {...props}
    type='checkbox'
    label={
      props.label ? (
        <PicassoForm.Label
          htmlFor={props.id}
          titleCase={props.titleCase}
          labelEndAdornment={props.labelEndAdornment}
        >
          {props.label}
        </PicassoForm.Label>
      ) : null
    }
  >
    {({
      // omitting highlight from inputProps
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      highlight,
      ...inputProps
    }: SwitchProps & { highlight?: 'autofill' }) => {
      return <PicassoSwitch {...inputProps} />
    }}
  </PicassoField>
)

Switch.defaultProps = {}

Switch.displayName = 'Switch'

export default Switch
