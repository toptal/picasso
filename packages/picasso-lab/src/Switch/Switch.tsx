import MUISwitch from '@material-ui/core/Switch'
import { makeStyles, Theme } from '@material-ui/core/styles'
import { BaseProps, TextLabelProps } from '@toptal/picasso-shared'
import React, { ButtonHTMLAttributes, forwardRef, ReactNode } from 'react'
import FormControlLabel from '@toptal/picasso/FormControlLabel'

import styles from './styles'

const useStyles = makeStyles<Theme, Props>(styles, { name: 'PicassoSwitch' })

export interface Props
  extends BaseProps,
    TextLabelProps,
    Omit<ButtonHTMLAttributes<HTMLButtonElement>, 'onChange'> {
  /** Show Switch initially as checked */
  checked?: boolean
  /** Disable changing `Switch` state */
  disabled?: boolean
  /** Text label for the `Switch` */
  label?: ReactNode
  /** Callback invoked when `Switch` changed its value */
  onChange?: (
    event: React.ChangeEvent<HTMLInputElement>,
    checked: boolean
  ) => void
  /** Value of the `Switch` (applicable only for controlled component) */
  value?: string
}

export const Switch = forwardRef<HTMLButtonElement, Props>(function Switch(
  props,
  ref
) {
  const {
    label,
    id,
    className,
    style,
    disabled,
    onChange,
    value,
    checked,
    titleCase,
    color, // eslint-disable-line
    'data-testid': dataTestId,
    ...rest
  } = props

  const classes = useStyles(props)

  const switchElement = (
    <MUISwitch
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...rest}
      color='primary'
      ref={ref}
      checked={checked}
      className={className}
      style={style}
      disabled={disabled}
      id={id}
      onChange={onChange}
      value={value}
      data-testid={label ? undefined : dataTestId}
    />
  )

  if (!label) {
    return switchElement
  }

  return (
    <FormControlLabel
      classes={{
        label: classes.label
      }}
      control={switchElement}
      disabled={disabled}
      label={label}
      titleCase={titleCase}
      data-testid={dataTestId}
    />
  )
})

Switch.defaultProps = {
  disabled: false,
  onChange: () => {}
}

Switch.displayName = 'Switch'

export default Switch
