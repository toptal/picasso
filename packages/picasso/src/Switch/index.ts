import type { OmitInternalProps } from '@toptal/picasso-shared'

import type { Props as InternalSwitchProps } from './Switch'
export { default } from './Switch'
export type SwitchProps = OmitInternalProps<InternalSwitchProps>
/** @deprecated [FX-4714] Use SwitchProps instead */
export type Props = SwitchProps
