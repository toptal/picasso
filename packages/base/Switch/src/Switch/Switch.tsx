import { Switch as MUISwitch } from '@mui/base/Switch'
import type { BaseProps, TextLabelProps } from '@toptal/picasso-shared'
import type { ButtonHTMLAttributes, ReactNode } from 'react'
import React, { forwardRef } from 'react'
import { FormControlLabel } from '@toptal/picasso-form'
import cx from 'classnames'

export interface Props
  extends BaseProps,
    TextLabelProps,
    Omit<ButtonHTMLAttributes<HTMLButtonElement>, 'onChange' | 'type'> {
  /** Show Switch initially as checked */
  checked?: boolean
  /** Disable changing `Switch` state */
  disabled?: boolean
  /** Text label for the `Switch` */
  label?: ReactNode
  /** Callback invoked when `Switch` changed its value */
  onChange?: (
    event: React.ChangeEvent<HTMLInputElement>,
    checked: boolean
  ) => void
}

export const Switch = forwardRef<HTMLButtonElement, Props>(function Switch(
  props,
  ref
) {
  const {
    label,
    id,
    className,
    style,
    disabled,
    onChange,
    checked,
    titleCase,
    color, // eslint-disable-line
    'data-testid': dataTestId,
    ...rest
  } = props

  const onChangeCallback: React.ChangeEventHandler<
    HTMLInputElement
  > = event => {
    if (onChange) {
      onChange(event, event.target.checked)
    }
  }

  const switchElement = (
    <MUISwitch
      {...rest}
      color='primary'
      ref={ref}
      checked={checked}
      className={className}
      style={style}
      disabled={disabled}
      id={id}
      onChange={onChangeCallback}
      data-testid={label ? undefined : dataTestId}
      slotProps={{
        root: {
          className:
            'w-[40px] h-[24px] p-0 relative inline-flex z-0 overflow-visible shrink-0 align-middle group',
        },
        track: {
          className: cx(
            'w-full h-full border border-solid bg-gray-600 border-gray-600 opacity-100 rounded-[12px]',
            'transition-colors duration-300 ease-out',
            'group-[.base--checked]:bg-blue-500 group-[.base--checked]:border-blue-500',
            'group-[.base--disabled]:opacity-40',
            'group-[.base--disabled:not(.base--checked)]:bg-black'
          ),
        },
        thumb: {
          className: cx(
            'w-[22px] h-[22px] bg-current text-white block rounded-full shadow-1 absolute z-10 p-0 top-[1px] left-[1px]',
            'transition-transform duration-150 ease-out',
            'group-[:not(.base--disabled):hover]:shadow-[0_0_0_4px_rgba(32,78,207,0.48)]',
            'group-[.base--focusVisible]:shadow-[0_0_0_4px_rgba(32,78,207,0.48)]',
            'group-[.base--checked]:translate-x-[16px]'
          ),
        },
        input: {
          className: cx(
            'w-[200%] h-full m-0 p-0 opacity-0 absolute top-0 -left-[50%] cursor-pointer z-20',
            'group-[.base--disabled]:cursor-default'
          ),
        },
      }}
    />
  )

  if (!label) {
    return switchElement
  }

  return (
    <FormControlLabel
      classes={{
        root: 'items-start text-lg',
        label: 'ml-[0.5em] mt-[0.25em] max-w-[calc(100%-1em-0.5em+1px)]',
      }}
      control={switchElement}
      disabled={disabled}
      label={label}
      titleCase={titleCase}
      data-testid={dataTestId}
    />
  )
})

Switch.defaultProps = {
  disabled: false,
  onChange: () => {},
}

Switch.displayName = 'Switch'

export default Switch
