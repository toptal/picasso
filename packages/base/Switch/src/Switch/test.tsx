import React from 'react'
import type { PicassoConfig } from '@toptal/picasso-test-utils'
import { render, fireEvent } from '@toptal/picasso-test-utils'
import type { OmitInternalProps } from '@toptal/picasso-shared'

import type { Props } from './Switch'
import { Switch } from './Switch'

const renderSwitch = (
  props: OmitInternalProps<Props>,
  picassoConfig?: PicassoConfig
) => {
  const { label, disabled, titleCase, onChange } = props

  return render(
    <Switch
      disabled={disabled}
      label={label}
      onChange={onChange}
      titleCase={titleCase}
      data-testid='switch'
    />,
    undefined,
    picassoConfig
  )
}

describe('Switch', () => {
  it('renders default Switch without label', () => {
    const { container } = renderSwitch({})

    expect(container).toMatchSnapshot()
  })

  it('renders default Switch with label', () => {
    const { getByTestId } = renderSwitch({ label: 'A Switch' })

    expect(getByTestId('switch')).toHaveTextContent('A Switch')
  })

  it('renders disabled state', () => {
    const { getByTestId } = renderSwitch({
      disabled: true,
      label: 'Disabled',
    })

    expect(getByTestId('switch')).toMatchSnapshot()
  })

  it('transforms text to title case when Picasso titleCase property is true', () => {
    const { getByTestId } = renderSwitch(
      { label: 'abc ac4' },
      { titleCase: true }
    )

    expect(getByTestId('switch')).toHaveTextContent('Abc Ac4')
  })

  it('transforms text to title case when Picasso titleCase property is true but the component property overrides it', () => {
    const { getByTestId } = renderSwitch(
      { label: 'abc dp3', titleCase: false },
      { titleCase: true }
    )

    expect(getByTestId('switch')).toHaveTextContent('abc dp3')
  })

  it('behaves correctly when interacting', () => {
    const onChange = jest.fn()
    const label = 'Switch'

    const { getByTestId } = renderSwitch({ onChange, label })

    fireEvent.click(getByTestId('switch'))

    expect(onChange).toHaveBeenCalled()
    expect(getByTestId('switch')).toMatchSnapshot()
  })
})
