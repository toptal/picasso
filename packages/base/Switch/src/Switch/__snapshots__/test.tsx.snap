// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch behaves correctly when interacting 1`] = `
<label
  class="PicassoFormControlLabel-root items-start text-lg"
  data-testid="switch"
>
  <span
    class="base-Switch base- w-[40px] h-[24px] p-0 relative inline-flex z-0 overflow-visible shrink-0 align-middle group"
    color="primary"
  >
    <span
      class="base-Switch w-full h-full border border-solid bg-gray border-gray opacity-100 rounded-[12px] transition-colors duration-300 ease-out group-[.base--checked]:bg-blue group-[.base--checked]:border-blue group-[.base--disabled]:opacity group-[.base--disabled:not(.base--checked)]:bg"
    />
    <span
      class="base-Switch w-[22px] h-[22px] bg-current text-white block rounded-full shadow-1 absolute z-10 p-0 top-[1px] left-[1px] transition-transform duration-150 ease-out group-[:not(.base--disabled):hover]:shadow group-[.base--focusVisible]:shadow group-[.base--checked]:translate-x"
    />
    <input
      class="base-Switch w-[200%] h-full m-0 p-0 opacity-0 absolute top-0 -left cursor-pointer z-20 group-[.base--disabled]:cursor"
      role="switch"
      type="checkbox"
    />
  </span>
  <span
    class="PicassoFormLabel-root PicassoFormLabel-inline PicassoFormControlLabel-label ml-[0.5em] mt-[0.25em] max-w-[calc(100%-1em"
  >
    <span
      class="PicassoFormLabel-medium"
    >
      Switch
    </span>
  </span>
</label>
`;

exports[`Switch renders default Switch without label 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <span
      class="base-Switch w-[40px] h-[24px] p-0 relative inline-flex z-0 overflow-visible shrink-0 align-middle group"
      color="primary"
      data-testid="switch"
    >
      <span
        class="base-Switch w-full h-full border border-solid bg-gray border-gray opacity-100 rounded-[12px] transition-colors duration-300 ease-out group-[.base--checked]:bg-blue group-[.base--checked]:border-blue group-[.base--disabled]:opacity group-[.base--disabled:not(.base--checked)]:bg"
      />
      <span
        class="base-Switch w-[22px] h-[22px] bg-current text-white block rounded-full shadow-1 absolute z-10 p-0 top-[1px] left-[1px] transition-transform duration-150 ease-out group-[:not(.base--disabled):hover]:shadow group-[.base--focusVisible]:shadow group-[.base--checked]:translate-x"
      />
      <input
        class="base-Switch w-[200%] h-full m-0 p-0 opacity-0 absolute top-0 -left cursor-pointer z-20 group-[.base--disabled]:cursor"
        role="switch"
        type="checkbox"
      />
    </span>
  </div>
</div>
`;

exports[`Switch renders disabled state 1`] = `
<label
  class="PicassoFormControlLabel-root items-start text-lg PicassoFormControlLabel-disabled"
  data-testid="switch"
>
  <span
    class="base-Switch base- w-[40px] h-[24px] p-0 relative inline-flex z-0 overflow-visible shrink-0 align-middle group"
    color="primary"
  >
    <span
      class="base-Switch w-full h-full border border-solid bg-gray border-gray opacity-100 rounded-[12px] transition-colors duration-300 ease-out group-[.base--checked]:bg-blue group-[.base--checked]:border-blue group-[.base--disabled]:opacity group-[.base--disabled:not(.base--checked)]:bg"
    />
    <span
      class="base-Switch w-[22px] h-[22px] bg-current text-white block rounded-full shadow-1 absolute z-10 p-0 top-[1px] left-[1px] transition-transform duration-150 ease-out group-[:not(.base--disabled):hover]:shadow group-[.base--focusVisible]:shadow group-[.base--checked]:translate-x"
    />
    <input
      class="base-Switch w-[200%] h-full m-0 p-0 opacity-0 absolute top-0 -left cursor-pointer z-20 group-[.base--disabled]:cursor"
      disabled=""
      role="switch"
      type="checkbox"
    />
  </span>
  <span
    class="PicassoFormLabel-root PicassoFormLabel-disabled PicassoFormLabel-inline PicassoFormControlLabel-label ml-[0.5em] mt-[0.25em] max-w-[calc(100%-1em"
  >
    <span
      class="PicassoFormLabel-medium"
    >
      Disabled
    </span>
  </span>
</label>
`;
